dvc/fs/azure.py

```diff
@@ -1,58 +1,94 @@
     REQUIRES = {
-        "azure-storage-blob": "azure.storage.blob",
+        "adlfs": "adlfs",
         "knack": "knack",
+        "azure-identity": "azure.identity",
     }
-    PARAM_CHECKSUM = "etag"
-    DETAIL_FIELDS = frozenset(("etag", "size"))
-
-    COPY_POLL_SECONDS = 5
-    LIST_OBJECT_PAGE_SIZE = 5000
 
     def __init__(self, repo, config):
         super().__init__(repo, config)
 
-        url = config.get("url", "azure://")
+        url = config.get("url")
         self.path_info = self.PATH_CLS(url)
         self.bucket = self.path_info.bucket
 
         if not self.bucket:
             container = self._az_config.get("storage", "container_name", None)
-            self.path_info = self.PATH_CLS(f"azure://{container}")
-            self.bucket = self.path_info.bucket
+            url = f"azure://{container}"
 
-        self._conn_str = config.get(
-            "connection_string"
-        ) or self._az_config.get("storage", "connection_string", None)
+        self.path_info = self.PATH_CLS(url)
+        self.bucket = self.path_info.bucket
+
+        self.login_method, self.login_info = self._prepare_credentials(config)
 
-        self._account_url = None
-        if not self._conn_str:
-            name = self._az_config.get("storage", "account", None)
-            self._account_url = f"https://{name}.blob.core.windows.net"
+    def _prepare_credentials(self, config):
+        from azure.identity.aio import DefaultAzureCredential
 
-        self._credential = config.get("sas_token") or self._az_config.get(
-            "storage", "sas_token", None
+        login_info = {}
+        login_info["connection_string"] = config.get(
+            "connection_string",
+            self._az_config.get("storage", "connection_string", None),
+        )
+        login_info["account_name"] = config.get(
+            "account_name", self._az_config.get("storage", "account", None)
+        )
+        login_info["account_key"] = config.get(
+            "account_key", self._az_config.get("storage", "key", None)
+        )
+        login_info["sas_token"] = config.get(
+            "sas_token", self._az_config.get("storage", "sas_token", None)
         )
-        if not self._credential:
-            self._credential = self._az_config.get("storage", "key", None)
+        login_info["tenant_id"] = config.get("tenant_id")
+        login_info["client_id"] = config.get("client_id")
+        login_info["client_secret"] = config.get("client_secret")
+
+        if not any(login_info.values()):
+            login_info["credentials"] = DefaultAzureCredential()
+
+        for login_method, required_keys in [  # noqa
+            ("connection string", ["connection_string"]),
+            ("account key", ["account_name", "account_key"]),
+            ("SAS token", ["account_name", "sas_token"]),
+            ("anonymous login", ["account_name"]),
+            (f"default credentials ({_DEFAULT_CREDS_STEPS})", ["credentials"]),
```

Comment:
@shcheklein would this work for showing user which method we tried (this order comes from the adlfs)? I couldn't figure out how else I can describe default credentials, so I just appended the documentation link which includes all the stages that meant to be tried out.

Reply:
look fine to me. I'm not sure about names and the order, etc - I would need to go too deep into details to say something meaningful re the details

Reply:
I don't see though our own credentials -when we take principal things like token_id etc and pass into adlfs, or when we read and parse config. Is is part of the `default credentials` in this case?

Reply:
The `default credentials` happen when you don't specify anything in your .dvc/config / ~/.azure/config or in the appropriate environment variables (AZURE_STORAGE_* stuff). If we can't find anything in those, we fall back to default credentials which covers stuff like AZURE_* environment variables etc. 

> look fine to me. I'm not sure about names and the order, etc - I would need to go too deep into details to say something meaningful re the details

There is not much about the order we can do, though I was wondering what do you think about the `login_method` and the AzureAuthError message we raise below

Reply:
yep, but I don't see in the list principal service auth defined in .dvc/config - client_id, tenant_id, etc ... is it intentional?

Reply:
Ah my bad🤦🏻! I forgot to add these options, thanks for catching it up

Reply:
resolved, but I don't see the change? forgot to push?